### PR TITLE
TriaIterator: return a proper Accessor temporary

### LINKED ISSUE #15440: Binding to TriaAccessor reference returned by dereferencing a TriaIterator temporary leads to use after scope
I see a segmentation fault in release mode in the test `matrix_free/pbc_orientation_02.release`:
```
0x00005555555a8942 in void dealii::MatrixFreeTools::categorize_by_boundary_ids<3, dealii::MatrixFree<3, double, dealii::VectorizedArray<double, 2ul> >::AdditionalData>(dealii::Triangulation<3, 3> const&, dealii::MatrixFree<3, double, dealii::VectorizedArray<double, 2ul> >::AdditionalData&) ()
(gdb) bt
#0  0x00005555555a8942 in void dealii::MatrixFreeTools::categorize_by_boundary_ids<3, dealii::MatrixFree<3, double, dealii::VectorizedArray<double, 2ul> >::AdditionalData>(dealii::Triangulation<3, 3> const&, dealii::MatrixFree<3, double, dealii::VectorizedArray<double, 2ul> >::AdditionalData&) ()
#1  0x00005555555cbdc0 in void test<3, 3>() ()
#2  0x00005555555cb679 in main ()
```
I am working on producing a better stack trace. Related failing tests (failing in optimized release mode)
```
matrix_free/categorize_by_boundary_ids_01.release
matrix_free/ecl_01.mpirun={1,3,5}.release
matrix_free/pbc_orientation_02.release
```

In reference to #15383

### PR DESCRIPTION
**Pull request by tamiko**
Closes #15440

### CODE REVIEW

File: include/deal.II/grid/tria_iterator.h [hunk 1]
```diff
@@ -1036,7 +1014,7 @@ template <typename Accessor>
 inline const Accessor *
 TriaRawIterator<Accessor>::operator->() const
 {
-  return &(this->operator*());
```
Comment (drwells) [inline include/deal.II/grid/tria_iterator.h:1039]: !!! How did we never notice this? !!!

**Comment by tamiko** [comment]
@bangerth The changes needed to fix this issue a fairly minimal.

File: include/deal.II/grid/tria_iterator.h (hunk 1 above)
Comment (tamiko) [reply include/deal.II/grid/tria_iterator.h:1039]: Two reasons I think. First, deal.II has its own idiosynchrasies and users tend to follow them (for example our `cell->`, `face->` usage throughout the examples, etc.). Secondly, for a long time it wasn't easily possible to take a reference of just about anything. This only came with `auto`. And even though this is C++ 11, the actual era of `auto` in the library started much later.

Comment (bangerth) [reply include/deal.II/grid/tria_iterator.h:1039]: But this isn't wrong. In fact, you're replacing the code with something I consider worse. The previous syntax made clear that `operator->` and `operator*` are at the core the same thing: One is implemented in terms of the other.

File: include/deal.II/grid/tria_iterator.h [hunk 2]
```diff
@@ -1036,7 +1014,7 @@ template <typename Accessor>
...
+  return &accessor;
```
Comment (bangerth) [inline include/deal.II/grid/tria_iterator.h:1017]: This does not actually improve anything. `operator->` still returns a pointer to a sub-object, and if that is a sub-object to a temporary, you're still screwed. So while
```
  auto face = *cell->face(f);
```
actually works,
```
  auto face = cell->face(f)->operator->();
```
does not. I'm not sure that's such a big improvement.

**Comment by bangerth** [comment]
I am not in favor of this patch at this point late in the game. There is exactly one place in the library where we have `*cell->`, and #15444 fixed it. Let us apply that patch.

I don't like this patch because it returns a copy of an object that consists of two pointers and two integers, a total of 24 bytes that I am certain happens millions of times in even small tutorial programs. I could be convinced that compilers can look through this with some timing data, but the day before we want to branch for a release is not the time for this -- in particular because I don't see evidence that that has led to frequent bugs: The issue in #15440 is, I believe, the first time in 25 years that this has caused an issue, and it took three people half an afternoon to figure it out and understand the root cause.

File: include/deal.II/grid/tria_iterator.h (hunk 2 above)
Comment (tamiko) [reply include/deal.II/grid/tria_iterator.h:1017]: @bangerth But `operator->` is actually accessing the underlying object while the temporary is alive,
 while `operator*` returns a derferenced object. There is a reason these are two different operations.

**Comment by tamiko** [comment]
@bangerth No one is talking about adding this patch to the 9.5 release. It is marked as a "draft" and it is added to the 9.6 milestone.

**Comment by tamiko** [comment]
@bangerth As requested here is some performance instrumentation with `step-6`: I have modified step-6 to do 2 additional global refinement steps and reduced the solver tolerance to 1e-6 in order to emphasize triangulation and iterator/accessor related operations a bit more.

```
                                         +-----------------------------------------+ +-----------------------------------------+
                                         |                   old                   | |                    new                  |
+------------------------------+---------+-------------+-------------+-------------+ +-------------+-------------+-------------+
|             Event            | Counter |  HWThread 5 |  HWThread 5 |  HWThread 5 | |  HWThread 5 |  HWThread 5 |  HWThread 5 |
+------------------------------+---------+-------------+-------------+-------------+ +-------------+-------------+-------------+
|       INSTR_RETIRED_ANY      |  FIXC0  | 54796301644 | 54796288129 | 54796304083 | | 54794572544 | 54794562292 | 54794564309 |
|     CPU_CLK_UNHALTED_CORE    |  FIXC1  | 27867832253 | 28118964055 | 28246917724 | | 28567170277 | 28462659981 | 28493449844 |
|     CPU_CLK_UNHALTED_REF     |  FIXC2  | 15858667104 | 15993406380 | 16102871568 | | 16245780096 | 16194584112 | 16213704780 |
| BR_INST_RETIRED_ALL_BRANCHES |   PMC0  |  7201294421 |  7201289621 |  7201295226 | |  7200769330 |  7200765659 |  7200765944 |
| BR_MISP_RETIRED_ALL_BRANCHES |   PMC1  |   157499884 |   158055341 |   157416224 | |   189713720 |   189726186 |   189588129 |
+------------------------------+---------+-------------+-------------+-------------+ +-------------+-------------+-------------+
                                         |                                         | |                                         |
                                         |                   old                   | |                    new                  |
            +----------------------------+-------------+-------------+-------------+ +-------------+-------------+-------------+
            |           Metric           |  HWThread 5 |  HWThread 5 |  HWThread 5 | |  HWThread 5 |  HWThread 5 |  HWThread 5 |
            +----------------------------+-------------+-------------+-------------+ +-------------+-------------+-------------+
            |     Runtime (RDTSC) [s]    |      7.7878 |      7.8466 |      7.9209 | |      7.9242 |      7.9242 |      7.9364 |
            |    Runtime unhalted [s]    |     13.2713 |     13.3900 |     13.4509 | |     13.5537 |     13.5537 |     13.5683 |
            |         Clock [MHz]        |   3690.0017 |   3692.1194 |   3683.7227 | |   3690.8340 |   3690.8340 |   3690.4669 |
            |             CPI            |      0.5086 |      0.5132 |      0.5155 | |      0.5194 |      0.5194 |      0.5200 |
            |         Branch rate        |      0.1314 |      0.1314 |      0.1314 | |      0.1314 |      0.1314 |      0.1314 |
            |  Branch misprediction rate |      0.0029 |      0.0029 |      0.0029 | |      0.0035 |      0.0035 |      0.0035 |
            | Branch misprediction ratio |      0.0219 |      0.0219 |      0.0219 | |      0.0263 |      0.0263 |      0.0263 |
            |   Instructions per branch  |      7.6092 |      7.6092 |      7.6092 | |      7.6095 |      7.6095 |      7.6095 |
            +----------------------------+-------------+-------------+-------------+ +-------------+-------------+-------------+
```
I do see a reduction in retired instruction cycles by a very small amount. However this particular example step suffers from a slightly higher branch misprediction rate of 0.0035 instead of 0.0029 and with this to an increase in runtime of less than 1%.

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
This looks to be in the noise.

I'm ok with this. I'd like it if @kronbichler looked at the numbers as well. He's got more experience with this kind of thing than I do.

**Review by kronbichler — APPROVED** [review]

**Comment by kronbichler** [comment]
```
/jenkins/workspace/dealii-serial_PR-15445/tests/matrix_free/pbc_orientation_01.cc:123:7: error: cannot bind non-const lvalue reference of type ‘dealii::CellAccessor<3, 3>&’ to an rvalue of type ‘dealii::CellAccessor<3, 3>’
  123 |       for (auto &cell : tria)
```
and a similar one in `pbc_orientation_02.cc`.

**Comment by bangerth** [comment]
That's an interesting failure. I'll fix this in a separate PR since it happens in a couple of other places as well.

**Pull request merged**